### ISSUE TITLE
None Grad with Custom Loss

### ISSUE THREAD

**Issue by lucabergamini** [pytorch/pytorch#2774]
Greetings, I'm trying a custom loss to minimize the IoU (intersection over union) between two rectangles, giving the coords of the four vertexes for both of them. What I'm trying to do for the single rectangle is basically to create a score map with 1 if the point is inside the bounds and 0 otherwise, where the points came from a grid 224x224 (I'm working on imagenet based networks). From there i can easily compute the intersection and union of two rectangles simply by sum and product of the masks. The forward part works smoothly, and so does the backward. However, even if I don't get any error, the grads are None everywhere, thus the network does not learn anything. Here is a minimum example which shows the problem (mind that replacing my loss with the MSE works as expected). I guess the gradient is lost somewhere, but I can't find where..

```python

import torch
import numpy
from torch.autograd import Variable
from torch.nn import MSELoss

from torch.nn.modules.loss import _Loss
from matplotlib import pyplot



class IoU_real(_Loss):
    def __init__(self):
        super(_Loss, self).__init__()
        # define the grid
        bounds = (224, 224)
        grid = numpy.mgrid[0:bounds[0], 0:bounds[1]]
        self.grid = numpy.concatenate((grid[0][..., numpy.newaxis], grid[1][..., numpy.newaxis]), axis=-1)
        self.grid = Variable(torch.FloatTensor(self.grid).cuda(), requires_grad=True)

    def _get_mask(self, rect):
        """
        get the rectangle area on the grid as a 2-values mask
        :param rect: 4x2 rect coords
        :return: 224x224 image with 2-values (0 stands for outside, 1 for inside)
        """


        #get coefficients of the 4 lines which limit the rectangle area
        coeff_x = []
        coeff_y = []
        coeff_q = []
        for i in xrange(4):
            if i != 3:
                p_1 = rect[i]
                p_2 = rect[i + 1]
            else:
                p_1 = rect[i]
                p_2 = rect[0]

            coeff_x.append(p_1[1] - p_2[1])
            coeff_y.append(p_2[0] - p_1[0])
            coeff_q.append(p_1[0] * p_2[1] - p_2[0] * p_1[1])

        coeff_x = torch.cat(coeff_x)
        coeff_y = torch.cat(coeff_y)
        coeff_q = torch.cat(coeff_q)

        #test every point in the grid against the equations
        # 224*224*1-> 224*224*4 -> 224*224*4
        score = self.grid[..., 0:1] * coeff_x + self.grid[..., 1:2] * coeff_y + coeff_q
        #now intersect the area
        score_final = Variable(torch.ones(list(score.size())[:-1]).cuda(), requires_grad=True)
        for i in xrange(4):
            score_final = score_final * (score[..., i] <= 0).float()
        #return the map
        return score_final

    def per_element_loss(self, out, label):
        out = torch.squeeze(out)
        label = torch.squeeze(label)
        #get both maskes
        m_1 = self._get_mask(out)
        m_2 = self._get_mask(label)
        #show them for debug
        pyplot.imshow((m_1 + m_2).data.cpu().numpy(), cmap="Greys")
        pyplot.show()
        #intersection is where the two cohexists
        intersection = torch.sum(m_1 * m_2)
        #union is the sum minus the intersection, otherwise it would be counted twice
        union = torch.sum(m_1 + m_2) - intersection
        #small fuzzy
        iou = intersection / (union + 10e-08)
        return iou

    def forward(self, outs, labels):
        """
        :param outs: out vertices as  n_el*1*4*2 
        :param labels: target vertices as  n_el*1*4*2
        :return: iou  
        """
        #define the accumulator
        iou_final = Variable(torch.FloatTensor(numpy.zeros(1)).cuda(), requires_grad=True)
        #iterate over couples
        for (out, label) in zip(outs, labels):
            # the metric is in fact 1-IoU (because we wat to minimize the loss)
            iou_final = iou_final + (1 - self.per_element_loss(out, label))
        #average
        return iou_final / len(outs)


if __name__ == "__main__":
    #two rectangles
    input = numpy.array(((50, 50), (50, 100), (100, 100), (100, 50)))
    target = numpy.array(((50, 50), (50, 100), (100, 100), (100, 50)))
    #torch
    input = Variable(torch.FloatTensor(input)).cuda()
    target = Variable(torch.FloatTensor(target)).cuda()
    target = target.view(1, -1)
    #we just add a bias
    bias = Variable(torch.ones(8).cuda() * 10, requires_grad=True)
    #instantiate the loss
    #MSE works..
    #loss = MSELoss().cuda()
    loss = IoU_real().cuda()

    input = input.view(1, -1)
    input = input + bias
    #compute the output
    res = loss(input.view(1, 1, 4, 2), target.view(1, 1, 4, 2) + 20)
    #IoU is not 0, so we should have a gradient wrt bias
    print res
    #do the backprop
    res.backward()
    #print the gradient
    print bias.grad




```

The code shows the two rectangles overlayed, and the IoU Loss is not 1, so it should propagate some gradient back..

![figure_1](https://user-images.githubusercontent.com/27865235/30541047-ee70764e-9c79-11e7-8c94-894585b7f8f4.png)

![figure_2](https://user-images.githubusercontent.com/27865235/30541066-077c743a-9c7a-11e7-9942-7634eb461ec6.png)

**Comment by lucabergamini** [comment]
This line seems the one which interferes with the gradient:
```
            score_final = score_final * (score[..., i] <= 0).float()
```

**Comment by SsnL** [comment]
If I understand correctly, the map from `_get_mask` is not differentiable. So the grad shouldn't exist at all. Maybe starts with a softened (and differentiable) loss, then gradually un-soften it?
